core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,306 @@
+        sendQueries(questions, hostname, result);
+        result.handle((entry, cause) -> {
+            if (cause != null) {
+                cache.remove(hostname);
+                promise.tryFailure(cause);
+                return null;
+            }
+            promise.trySuccess(new InetSocketAddress(entry.address(), port));
+            return null;
+        });
+    }
+
+    private static void handleFromCache(CompletableFuture<CacheEntry> future,
+                                        Promise<InetSocketAddress> promise, int port) {
+        future.handle((entry, cause) -> {
+            if (cause != null) {
+                promise.tryFailure(cause);
+                return null;
+            }
+            entry.servedFromCache();
+            promise.trySuccess(new InetSocketAddress(entry.address(), port));
+            return null;
+        });
+    }
+
+    private void sendQueries(List<DnsQuestion> questions, String hostname,
+                             CompletableFuture<CacheEntry> result) {
+        final Future<List<DnsRecord>> recordsFuture = resolver.sendQueries(questions, hostname);
+        recordsFuture.addListener(f -> {
+            if (!f.isSuccess()) {
+                result.completeExceptionally(f.cause());
+                return;
+            }
+
+            @SuppressWarnings("unchecked")
+            final List<DnsRecord> records = (List<DnsRecord>) f.getNow();
+            InetAddress inetAddress = null;
+            long ttlMillis = -1;
+            try {
+                for (DnsRecord r : records) {
+                    final byte[] addrBytes = extractAddressBytes(r, logger, hostname);
+                    if (addrBytes == null) {
+                        continue;
+                    }
+                    try {
+                        inetAddress = InetAddress.getByAddress(hostname, addrBytes);
+                        ttlMillis = TimeUnit.SECONDS.toMillis(
+                                Math.max(Math.min(r.timeToLive(), maxTtl), minTtl));
+                        break;
+                    } catch (UnknownHostException e) {
+                        // Should never reach here because we already validated it in extractAddressBytes.
+                        result.completeExceptionally(
+                                new IllegalArgumentException("Invalid address: " + hostname, e));
+                        return;
+                    }
+                }
+            } finally {
+                records.forEach(ReferenceCountUtil::safeRelease);
+            }
+
+            if (inetAddress == null) {
+                result.completeExceptionally(
+                        new UnknownHostException("failed to receive DNS records for " + hostname));
+                return;
+            }
+            final CacheEntry entry = new CacheEntry(inetAddress, ttlMillis, questions);
+            result.complete(entry);
+            entry.scheduleRefresh(ttlMillis);
```

Comment:
nit: How about moving this logic up `records.forEach(ReferenceCountUtil::safeRelease)` to complete the future quickly?

Reply:
I didn't get what you mean. 😄

Reply:
I thought we can release `recodes` lazy... but I guess it is meaningless. Sorry for the confusion. 😀